webpack/react_app/components/TargetingHosts/index.js

```diff
@@ -1,37 +1,87 @@
-import React, { useEffect } from 'react';
+import React, { useEffect, useState } from 'react';
 import { useSelector, useDispatch } from 'react-redux';
+
+import { useForemanSettings } from 'foremanReact/Root/Context/ForemanContext';
 import { stopInterval } from 'foremanReact/redux/middlewares/IntervalMiddleware';
-import TargetingHosts from './TargetingHosts';
+import { STATUS } from 'foremanReact/constants';
 
 import {
   selectItems,
-  selectStatus,
+  selectApiStatus,
   selectAutoRefresh,
+  selectTotalHosts,
+  selectIntervalExists,
 } from './TargetingHostsSelectors';
+import { getApiUrl } from './TargetingHostsHelpers';
 import { getData } from './TargetingHostsActions';
 import { TARGETING_HOSTS } from './TargetingHostsConsts';
+import TargetingHostsPage from './TargetingHostsPage';
 
 const WrappedTargetingHosts = () => {
   const dispatch = useDispatch();
+  const { perPage, perPageOptions } = useForemanSettings();
+
   const autoRefresh = useSelector(selectAutoRefresh);
   const items = useSelector(selectItems);
-  const status = useSelector(selectStatus);
+  const apiStatus = useSelector(selectApiStatus);
+  const totalHosts = useSelector(selectTotalHosts);
+  const [searchQuery, setSearchQuery] = useState('');
+  const [pagination, setPagination] = useState({
+    page: 1,
+    perPage,
+    perPageOptions,
+  });
+  const [apiUrl, setApiUrl] = useState(getApiUrl(searchQuery, pagination));
+  const intervalExists = useSelector(selectIntervalExists);
 
-  useEffect(() => {
-    dispatch(getData());
+  const handleSearch = query => {
+    const defaultPagination = { page: 1, perPage: pagination.perPage };
+    stopApiInterval();
 
-    return () => {
+    setApiUrl(getApiUrl(query, defaultPagination));
+    setSearchQuery(query);
+    setPagination(defaultPagination);
+  };
+
+  const handlePagination = args => {
+    stopApiInterval();
+    setPagination(args);
+    setApiUrl(getApiUrl(searchQuery, args));
+  };
+
+  const stopApiInterval = () => {
+    if (intervalExists) {
       dispatch(stopInterval(TARGETING_HOSTS));
-    };
-  }, [dispatch]);
+    }
+  };
 
   useEffect(() => {
+    dispatch(getData(apiUrl));
+
     if (autoRefresh === 'false') {
       dispatch(stopInterval(TARGETING_HOSTS));
     }
-  }, [autoRefresh, dispatch]);
 
-  return <TargetingHosts status={status} items={items} />;
+    return () => {
+      dispatch(stopInterval(TARGETING_HOSTS));
+    };
+  }, [dispatch, apiUrl, autoRefresh]);
+
+  if (apiStatus === STATUS.ERROR) {
+    stopApiInterval();
```

Comment:
you can use the new api middleware callbacks, instead of waiting for the state to update with the status
e.g: 
```js
get({ 
  key: TARGETING_HOSTS, 
  url,
  handleError: () => { stopInterval() } 
});
```

Reply:
It doesn't work for me, api call end up in never-ending loop.

Reply:
just make sure that you pass it with `dispatch` :)